fabtests/pytest/common.py

```diff
@@ -384,34 +394,58 @@ def run(self):
                               self._cmdline_args.strict_fabtests_mode)
 
 
-class MultinodeTest:
+class MultinodeTest(ClientServerTest):
 
-    def __init__(self, cmdline_args, base_command, numproc):
+    def __init__(self, cmdline_args, server_base_command,
+                 client_base_command, client_hostname_list,
+                 run_client_asynchronously=True):
         self._cmdline_args = cmdline_args
-        self._base_command = base_command
-        self._numproc = numproc
+        self.numclient = len(client_hostname_list)
         self._timeout = self._cmdline_args.timeout
 
-        multinode_command = self._base_command + " -n {}".format(self._numproc)
-        self._server_command = cmdline_args.populate_command(multinode_command, "server", self._timeout)
-        self._client_command = cmdline_args.populate_command(multinode_command, "client", self._timeout)
+        self._server_base_command = cmdline_args.populate_command(server_base_command, "server", self._timeout)
+        self._client_base_command_list = []
+        for client_hostname in client_hostname_list:
+            cmdline_args_copy = copy.copy(cmdline_args)
+            cmdline_args_copy.client_id = client_hostname
+            self._client_base_command_list.append(cmdline_args_copy.populate_command(client_base_command, "client", self._timeout))
+        self._run_client_asynchronously = run_client_asynchronously
 
+    @retry(retry_on_exception=is_ssh_connection_error, stop_max_attempt_number=3, wait_fixed=SERVER_RESTART_DELAY_MS)
     def run(self):
-        if self._cmdline_args.is_test_excluded(self._base_command):
+        if self._cmdline_args.is_test_excluded(self._server_base_command):
+            pytest.skip("excluded")
+
+        # _client_base_command_list is populated from the same client_base_command,
+        # so it should be safe to only check the first one.
+        if self._cmdline_args.is_test_excluded(self._client_base_command_list[0]):
             pytest.skip("excluded")
 
         server_outfile = NamedTemporaryFile(prefix="fabtests_server.out.").name
 
         # start running
-        server_process = Popen(self._server_command + "> " + server_outfile + " 2>&1", shell=True)
+        server_process = Popen(self._server_base_command + "> " + server_outfile + " 2>&1", shell=True)
         sleep(1)
 
-        numclient = self._numproc - 1
-        client_process_list = [None] * numclient
-        client_outfile_list = [None] * numclient
-        for i in range(numclient):
+        client_process_list = [None] * self.numclient
+        client_outfile_list = [None] * self.numclient
+        for i in range(self.numclient):
             client_outfile_list[i] = NamedTemporaryFile(prefix="fabtests_client_{}.out.".format(i)).name
-            client_process_list[i] = Popen(self._client_command + "> " + client_outfile_list[i] + " 2>&1", shell=True)
+            client_command = self._client_base_command_list[i]  + " 2>&1"
```

Comment:
`2>&1` is redundant because of `stderr=STDOUT`

Reply:
Thank you! Done